runners/direct-java/src/main/java/org/apache/beam/runners/direct/CopyOnAccessInMemoryStateInternals.java

```diff
@@ -355,6 +379,12 @@ private boolean containedInUnderlying(StateNamespace namespace, StateTag<? super
           public <KeyT, ValueT> MapState<KeyT, ValueT> bindMap(
               StateTag<? super K, MapState<KeyT, ValueT>> address,
               Coder<KeyT> mapKeyCoder, Coder<ValueT> mapValueCoder) {
+            if (mapKeyCoder == null || mapValueCoder == null) {
```

Comment:
So actually I am going to have to say "yes". Just like every PCollection needs a coder, every pieces of state needs a coder. The portability goals of Beam include a runner written in Java running a pipeline written in Python, and vice-versa. Data in the underlying model (the Runner API / Fn API) is always just bytes with a coder (spec) to interpret it.